### PR TITLE
Updated RELEASES.md for 1.21.0

### PR DESCRIPTION
**Pull request by XAMPPRocky**
[Rendered](https://github.com/Aaronepower/rust/blob/master/RELEASES.md)

### CODE REVIEW

File: RELEASES.md [hunk 1]
```diff
@@ -1,3 +1,61 @@
...
+  Example: `struct Foo<T>(T); let bar = Foo::<u8>(0);`.
+
+Compiler
+--------
+- [Upgraded jemalloc to 4.5.0][43911]
+- [Enabled unwinding panics on Redox][43917]
+
+Libraries
+---------
+- [Generate builtin impls for `Clone` for `[T: N]` where `N` is between 0
```
Comment (petrochenkov) [inline RELEASES.md]: typo `[T; N]`

File: RELEASES.md [hunk 2]
```diff
@@ -1,3 +1,61 @@
+Version 1.21.0 (2017-10-12)
+==========================
+
+Language
+--------
+- [Relaxed path syntax. You can now add type parameters to values][43540]
+  Example: `struct Foo<T>(T); let bar = Foo::<u8>(0);`.
```
Comment (petrochenkov) [inline RELEASES.md]: `Foo::<u8>(0)` was always possible, what https://github.com/rust-lang/rust/pull/43540 does is permitting (sometimes unnecessary) `::` before `<` in all paths, including paths passed to macros.
Previously:
```rust
my_macro!(Vec<i32>::new); // OK
my_macro!(Vec::<i32>::new); // ERROR
```
Now:
```rust
my_macro!(Vec<i32>::new); // OK
my_macro!(Vec::<i32>::new); // OK too
```

File: RELEASES.md [hunk 3]
```diff
@@ -1,3 +1,61 @@
...
+  to [doc.rust-lang.org/cargo](//doc.rust-lang.org/cargo)
+- [The rustdoc book is now available][43863]
+  at [doc.rust-lang.org/rustdoc](//doc.rust-lang.org/rustdoc)
+
+Compatibility Notes
+-------------------
+- [Remove the trait selection impl in method::probe][43880] This may cause
+  breakage in subtyping corner cases.
+- [make JSON error's byte position start at top of file.][42973] Was previously
+  relative to the `CodeMap` which required unstable code.
```
Comment (zackmdavis) [inline RELEASES.md]: Release note readers shouldn't need to think about named internal implementation details like `CodeMap`, and I'm not sure where "which required unstable code" is coming from.

Maybe something like "was previously relative to a concatenation of files in a crate"?? (I don't actually know how the compiler determines which files go in a `CodeMap`—and neither does typical consumer of error-message JSON, which is why we fixed this).

**Comment by zackmdavis** [comment]
There should also be a compatibility bullet point for [the `unused_results` lint getting stronger](https://github.com/rust-lang/rust/issues/44119). "`unused_results` lint no longer ignores booleans"?

File: RELEASES.md [hunk 4]
```diff
@@ -1,3 +1,61 @@
...
+----
+- [Cargo docs are moving][43916]
+  to [doc.rust-lang.org/cargo](//doc.rust-lang.org/cargo)
+- [The rustdoc book is now available][43863]
+  at [doc.rust-lang.org/rustdoc](//doc.rust-lang.org/rustdoc)
+
+Compatibility Notes
+-------------------
+- [Remove the trait selection impl in method::probe][43880] This may cause
+  breakage in subtyping corner cases.
```
Comment (kennytm) [inline RELEASES.md]: Is there any breakage examples?

BTW "trait selection impl in method::probe" sounds too technical for readers.

Comment (Aaronepower) [reply RELEASES.md]: It definitely is I just couldn't think of anything better. @arielb1 would you know a better explanation and example?

**Comment by Aaronepower** [comment]
@zackmdavis Could you provide a PR link? I can't seem to find the PR that makes the lint stronger.

File: RELEASES.md [hunk 5]
```diff
@@ -1,3 +1,65 @@
...
+Cargo
+-----
+- [You can now call `cargo install` with multiple package names][cargo/4216]
+- [Cargo commands inside a virtual workspace will now implicitly
+  pass `--all`][cargo/4335]
+
+Misc
+----
+- [Cargo docs are moving][43916]
+  to [doc.rust-lang.org/cargo](https://doc.rust-lang.org/cargo)
```
Comment (alexcrichton) [inline RELEASES.md]: We may not want to highlight this just yet as they're still a redirect to doc.crates.io, but soon though!

**Review by alexcrichton — COMMENTED** [review]
Thanks @Aaronepower!

I think this may also want to mention [`[patch]`](https://github.com/rust-lang/cargo/pull/4123) in Cargo (detailed in [RFC 1969](https://github.com/rust-lang/rfcs/pull/1969))

**Comment by alexcrichton** [comment]
Another Cargo change that may be worth mentioning is https://github.com/rust-lang/cargo/pull/4270, a change to the ignore/exclude syntax

**Comment by alexcrichton** [comment]
Also rust-lang/cargo#4400, the addition of the `--all-targets` flag

**Comment by alexcrichton** [comment]
Also rust-lang/cargo#4364, a new compatibility note

**Comment by alexcrichton** [comment]
Speaking of lint changes, I believe `unused_extern_crate` was turned on by default in https://github.com/rust-lang/rust/pull/42588

File: RELEASES.md (hunk 5 above)
Comment (Aaronepower) [reply RELEASES.md]: I didn't think so either, but it was marked as `relnotes` so I included it.

**Comment by zackmdavis** [comment]
@Aaronepower 

> Could you provide a PR link? I can't seem to find the PR that makes the lint stronger.

The change was, somewhat embarrassingly, unintentional in #43728 (removed a `TyBool => return` match arm without realizing that this would let us "fall through" to the part of the lint-pass that sets `unused_results`), but then the discussion and final-comment-period in #44119 retroactively ruled it a bugfix rather than a regression.

File: RELEASES.md [hunk 6]
```diff
@@ -1,3 +1,81 @@
...
+Compiler
+--------
+- [Upgraded jemalloc to 4.5.0][43911]
+- [Enabled unwinding panics on Redox][43917]
+- [Now emits a warning when you have an unused `extern crate`][42588]
+
+Libraries
+---------
+- [Generate builtin impls for `Clone` for `[T; N]` where `N` is between 0
+  and 32][43690]
```
Comment (ollie27) [inline RELEASES.md]: It's more than this. `Clone` is now implemented for all arrays with `T: Clone`, all tuples with every member being `Clone` and all function pointers.

File: RELEASES.md [hunk 7]
```diff
@@ -1,3 +1,81 @@
...
+- [You can now call `cargo install` with multiple package names][cargo/4216]
+- [Cargo commands inside a virtual workspace will now implicitly
+  pass `--all`][cargo/4335]
+- [Added a `[patch]` section to `Cargo.toml` to handle
+  prepublication dependencies][cargo/4123] [RFC 1969]
+- [`include` & `exclude` fields in `Cargo.toml` now accept gitignore
+  like patterns][cargo/4270]
+- [Added the `--all-targets` option][cargo/4400]
+- [Using required dependencies as a feature is now deprecated and emits
+  a warning][cargo/]
```
Comment (ollie27) [inline RELEASES.md]: missing link

**Review by ollie27 — COMMENTED** [review]
Looks like #43838 is missing.

File: RELEASES.md (hunk 4 above)
Comment (kennytm) [reply RELEASES.md]: @Aaronepower: Maybe summarize or just link to https://github.com/rust-lang/rust/issues/44224#issuecomment-330058902 ?

**Comment by alexcrichton** [comment]
gah sorry for letting this sit @Aaronepower! slipped off my radar :(

In any case looks great, thanks so much again for tackling this! Let's get this in and we can continue to iterate in-tree

@bors: r+

**Comment by kennytm** [comment]
@bors retry Wat.

Spurious network error, can't resolve crates.io.

**Comment by nrc** [comment]
Adds the RLS! `rustup component add rls-preview`

**Comment by Aaronepower** [comment]
@nrc I've added that, let know if the PR I picked is the wrong one to link.

File: RELEASES.md [hunk 8]
```diff
@@ -1,3 +1,94 @@
...
+  Example:
+  ```rust
+  my_macro!(Vec<i32>::new); // Always worked
+  my_macro!(Vec::<i32>::new); // Now works
+  ```
+- [You can now use static references for literals.][43838]
+  Example:
+  ```rust
+  fn main() {
+      let x: &'static u32 = 0;
```
Comment (kennytm) [inline RELEASES.md]: This should be `&0`, not `0`.

**Review by kennytm — CHANGES_REQUESTED** [review]

**Comment by dtolnay** [comment]
We are stabilizing std::mem::discriminant in 1.21. #44263,  #44552

We could backport any of the following library stabilizations -- I'm not clear how we decide whether to do that.

- TcpStream::connect_timeout #44563
- Iterator::for_each #44567 
- Ord::min/max #44593
- std::sync::atomic::compiler_fence #44595
- std::mem::needs_drop #44639
- Vec::splice #44640

**Comment by QuietMisdreavus** [comment]
The docs got a few notable quality-of-life additions. Not sure whether they're all "release notes" worthy, but i'll note them here all the same:

* [`std::os` docs for Unix, Linux, and Windows now appear in all docs, including the ones online](https://github.com/rust-lang/rust/pull/43348)
* [references](https://github.com/rust-lang/rust/pull/43560) and [function pointers](https://github.com/rust-lang/rust/pull/43529) have their own primitive pages in the std docs
* [trait implementors now print their associated types without having to go to the type's page to find them](https://github.com/rust-lang/rust/pull/43515)

**Comment by michaelwoerister** [comment]
I think the "async-llvm" changes from https://github.com/rust-lang/rust/pull/43506 will be in the next release (right, @alexcrichton?). This feature is [pretty neat](http://perf.rust-lang.org/?start=2017-07-30T00%3A00%3A00%2B00%3A00&end=2017-08-03T00%3A00%3A00%2B00%3A00&crates=%5B%22regex-0.1.80%40020-incr-from-scratch%22%2C%22syntex-0.42.2%40010-incr%22%5D&stat=max-rss) and it solves a fundamental problem we started running into with incremental compilation: Before, the amount of memory consumed by the LLVM part of compilation was determined by the total size of your crate. After, it is determined by the size of the largest N codegen-units, where N is the number of parallel threads you tell rustc to use.  Note that this is something especially important for incr. comp. but the feature helps with any setup that uses `codegen-units=N`.

Together with @alexcrichton's jobserver changes, this makes using something like `codegen-units= 2 * cpu-core-count` much more viable, if runtime performance is not the highest priority (e.g. most kinds of development builds).

File: RELEASES.md [hunk 9]
```diff
@@ -1,3 +1,94 @@
...
+  }
+  ```
+
+Compiler
+--------
+- [Upgraded jemalloc to 4.5.0][43911]
+- [Enabled unwinding panics on Redox][43917]
+- [Now emits a warning when you have an unused `extern crate`][42588]
```
Comment (arielb1) [inline RELEASES.md]: This is being backed out (#44825)

**Comment by alexcrichton** [comment]
Alright the release is just around the corner, so let's merge!

@bors: r+

**Pull request merged**